File: 16-3sum-closest/16-3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
       //base case
        int gap= Integer.MAX_VALUE,ans=0;
        if(nums.length<3)return 0;
        Arrays.sort(nums);
        int n=nums.length;
        for (int i = 0; i < n; i++) {
            int L=i+1,R=n-1;
            while (L<R){
                int curSum=nums[i]+nums[L]+nums[R];
                int curGap=Math.abs(curSum-target);
                if(curGap<gap){
                    gap=curGap;
                    ans=curSum;
                }
                if(curSum==target){
                    return target;
                } else if (curSum<target) {
                    L++;
                }else {
                    R--;
                }
            }
        }

        return ans;
    }
}

File: 16-3sum-closest/16-3sum-closest.ts
function threeSumClosest(nums: number[], target: number): number {
  let gap:number=Number.MAX_SAFE_INTEGER;
    let ans:number=0;
    nums.sort((a,b)=>a-b);
    for(let i=0;i<nums.length-2;i++){
        let j=i+1;
        let k=nums.length-1;
        while(j<k){
            let sum=nums[i]+nums[j]+nums[k];
            if(sum==target){
                return target;
            }
            if(Math.abs(sum-target)<gap){
                gap=Math.abs(sum-target);
                ans=sum;
            }
            if(sum<target){
                j++;
            }else{
                k--;
            }
        }
    }
    
    return ans;
};